// @colyseus/auth/build/JWT.mjs
// packages/auth/src/JWT.ts
import jsonwebtoken from "jsonwebtoken";
import { expressjwt } from "express-jwt";
var JWT = {
  settings: {
    /**
     * The secret used to sign and verify the JWTs.
     */
    secret: void 0,
    verify: {
      /**
       * The first algorithm in the list is used to sign new tokens.
       */
      algorithms: ["HS256"]
    }
  },
  sign: function(payload, options = {}) {
    return new Promise((resolve, reject) => {
      if (options.algorithm === void 0) {
        options.algorithm = JWT.settings.verify.algorithms[0];
      }
      jsonwebtoken.sign(payload, getJWTSecret(), options, (err, token) => {
        if (err) reject(err.message);
        resolve(token);
      });
    });
  },
  verify: function(token, options) {
    return new Promise((resolve, reject) => {
      jsonwebtoken.verify(token, getJWTSecret(), options || JWT.settings.verify, function(err, decoded) {
        if (err) reject(err);
        resolve(decoded);
      });
    });
  },
  /**
   * Returns the decoded payload without verifying if the signature is valid
   */
  decode: jsonwebtoken.decode,
  /**
   * Get express middleware that verifies JsonWebTokens and sets `req.auth`.
   */
  middleware: function(params) {
    return expressjwt(Object.assign({
      secret: getJWTSecret(),
      // credentialsRequired: false,
      algorithms: JWT.settings.verify.algorithms,
      ...JWT.settings.verify
    }, params));
  }
};
function getJWTSecret() {
  JWT.settings.secret ||= process.env.JWT_SECRET;
  if (!JWT.settings.secret) {
    console.error("\u274C Please provide 'JWT_SECRET' environment variable, or set 'JWT.settings.secret'.");
  }
  return JWT.settings.secret;
}
export {
  JWT
};
