// @colyseus/auth/build/oauth.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var oauth_exports = {};
__export(oauth_exports, {
  oAuthProviderCallback: () => oAuthProviderCallback,
  oauth: () => oauth
});
module.exports = __toCommonJS(oauth_exports);
var import_fs = __toESM(require("fs"));
var import_path = __toESM(require("path"));
var import_express = __toESM(require("express"));
var import_grant = __toESM(require("grant"));
var import_express_session = __toESM(require("express-session"));
var import_core = require("@colyseus/core");
var import_auth = require("./auth.js");
var import_connect_redis = __toESM(require("connect-redis"));
var import_JWT = require("./JWT.js");
let oAuthProviderCallback = async (data, provider) => {
  console.debug("OAuth callback missing. Use oauth.onCallback() to persist user data.");
  return data;
};
const oauth = {
  /**
   * Default 'grant' module configuration.
   */
  defaults: {
    // TODO: automatically retrieve "origin" based on "publicAddress" config [?]
    // origin: process.env.BACKEND_URL,
    transport: "session",
    state: true,
    response: ["tokens", "raw", "profile"]
  },
  /**
   * Route prefix for OAuth routes.
   */
  prefix: "/provider",
  providers: {},
  /**
   * Add OAuth provider configuration.
   * @param providerId OAuth provider name
   * @param config OAuth provider configuration
   */
  addProvider: function(providerId, config) {
    this.providers[providerId] = config;
  },
  /**
   * Provides a callback function that is called when OAuth is successful.
   */
  onCallback: function(callback) {
    oAuthProviderCallback = callback;
  },
  /**
   * Returns an Express Router that handles OAuth for configured providers.
   * @param callback (optional) Callback function that is called when OAuth is successful.
   * @returns Express Router
   */
  routes: function(callback) {
    if (callback) {
      this.onCallback(callback);
    }
    const router = import_express.default.Router();
    import_core.matchMaker.onReady.then(() => {
      const store = import_core.matchMaker.presence["pub"] ? new import_connect_redis.default({ client: import_core.matchMaker.presence["pub"] }) : void 0;
      const sessionMiddleware = (0, import_express_session.default)({
        secret: process.env.SESSION_SECRET,
        resave: false,
        saveUninitialized: false,
        // true
        store
      });
      const config = Object.assign({ defaults: this.defaults }, this.providers);
      config.defaults.prefix = oauth.prefix;
      router.use(sessionMiddleware);
      router.get("/:providerId", async (req, res, next) => {
        const providerId = req.params.providerId;
        if (oauth.providers[providerId]) {
          next();
        } else {
          if (process.env.NODE_ENV === "production") {
            res.send(`Missing OAuth provider configuration for "${providerId}".`);
          } else {
            const helpURLs = JSON.parse(import_fs.default.readFileSync(import_path.default.normalize(__dirname + "/../oauth_help_urls.json")).toString());
            const providerUrl = helpURLs[providerId];
            res.send(`<!doctype html>
<html>
<head>
<title>Missing "${providerId}" provider configuration</title>
<style>p { font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Noto Sans",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji"; }</style>
</head>
<body>
<p>Missing config for "${providerId}" OAuth provider.</p>
<hr />
<p><small><strong>Config example:</strong></small></p>
<pre><code>import { auth } from "@colyseus/auth";<br />
auth.oauth.addProvider("${providerId}", {
  key: "xxx",
  secret: "xxx",
});
</code></pre>
${providerUrl ? `<hr/><p><small><em>(Get your keys from <a href="${providerUrl}" target="_blank">${providerUrl}</a>)</em></small></p>` : ""}
</body>
</html>`);
          }
        }
      });
      router.use(import_grant.default.express(config));
      router.get("/:providerId/callback", async (req, res) => {
        const session2 = req.session;
        let user = null;
        let token = null;
        let response = void 0;
        if (session2.grant.response.error) {
          response = { error: session2.grant.response.error, user, token };
        } else {
          const data = session2.grant.response;
          if (session2.grant.dynamic?.token) {
            data.upgradingToken = await import_JWT.JWT.verify(session2.grant.dynamic?.token);
          }
          if (data.profile) {
            data.profile = oauth.transformProfileData(data.profile);
          }
          user = await oAuthProviderCallback(data, session2.grant.provider);
          token = await import_auth.auth.settings.onGenerateToken(user);
          response = { user, token };
        }
        res.send(`<!DOCTYPE html><html><head><script type="text/javascript">window.opener.postMessage(${JSON.stringify(response)}, '*');</script></head><body></body></html>`);
        res.end();
      });
    });
    return router;
  },
  /**
   * Transform raw profile data into a single object.
   * (e.g. Twitch returns an array of profiles, but we only need the first one)
   * @param raw
   */
  transformProfileData(raw) {
    if (raw.data && Array.isArray(raw.data) && raw.data.length === 1) {
      return raw.data[0];
    } else {
      return raw;
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  oAuthProviderCallback,
  oauth
});
